File: src/grove/dash/installer.py
"""Install/uninstall Grove hooks into Claude Code's settings.json."""

from __future__ import annotations

import json
import shutil
import sys
from datetime import datetime
from pathlib import Path

CLAUDE_SETTINGS = Path.home() / ".claude" / "settings.json"

# All hook event types Claude Code supports
HOOK_EVENTS = [
    "PreToolUse",
    "PostToolUse",
    "PostToolUseFailure",
    "Stop",
    "SessionStart",
    "SessionEnd",
    "Notification",
    "PermissionRequest",
    "UserPromptSubmit",
    "SubagentStart",
    "SubagentStop",
    "PreCompact",
    "TaskCompleted",
]

# Marker to identify our hooks
_GROVE_MARKER = "gw _hook"
# Legacy marker from pre-v0.13 (used sys.executable + python -m grove.dash).
# Needed so uninstall can clean up old hooks. Safe to remove in v0.14+.
_LEGACY_MARKER = "grove.dash"


def _hook_command(event: str) -> str:
    """Build the hook command string for a given event.

    Uses the ``gw`` console-script entry point (stable across upgrades)
    rather than the versioned Python interpreter path, which breaks when
    Homebrew or uv upgrades Grove to a new version.
    """
    gw = _resolve_gw()
    return f"GROVE_EVENT={event} {gw} _hook --event {event}"


def _resolve_gw() -> str:
    """Find the stable ``gw`` binary path.

    Prefers ``shutil.which`` so we get the canonical PATH entry (e.g.
    /opt/homebrew/bin/gw) rather than a version-specific Cellar path.
    Falls back to ``sys.executable -m grove.dash`` for editable/dev installs
    where ``gw`` might not be on PATH.
    """
    gw_path = shutil.which("gw")
    if gw_path:
        return gw_path
    # Fallback for dev installs — multi-token string, interpreted by shell
    return f"{sys.executable} -m grove.dash"


def _is_grove_hook(hook: dict) -> bool:
    """Check if a hook entry belongs to Grove."""
    cmd = hook.get("command", "")
    return _GROVE_MARKER in cmd or _LEGACY_MARKER in cmd


def install_hooks(dry_run: bool = False) -> dict[str, list[str]]:
    """Install Grove hooks into Claude Code settings.

    Returns a dict of {event: [action]} describing what was done.
    """
    actions: dict[str, list[str]] = {}

    # Load existing settings
    settings: dict = {}
    if CLAUDE_SETTINGS.exists():
        try:
            settings = json.loads(CLAUDE_SETTINGS.read_text())
        except (json.JSONDecodeError, OSError):
            settings = {}

    hooks = settings.setdefault("hooks", {})

    for event in HOOK_EVENTS:
        command = _hook_command(event)
        # Claude Code expects: {"matcher": "", "hooks": [{"type": "command", ...}]}
        rule_entry = {
            "matcher": "",
            "hooks": [{"type": "command", "command": command}],
        }

        event_hooks = hooks.get(event, [])
        if not isinstance(event_hooks, list):
            event_hooks = []

        # Find existing grove rule
        grove_idx = None
        for i, rule in enumerate(event_hooks):
            if not isinstance(rule, dict):
                continue
            for inner in rule.get("hooks", []):
                if isinstance(inner, dict) and _is_grove_hook(inner):
                    grove_idx = i
                    break
            if grove_idx is not None:
                break

        if grove_idx is not None:
            event_hooks[grove_idx] = rule_entry
            actions.setdefault(event, []).append("updated")
        else:
            event_hooks.append(rule_entry)
            actions.setdefault(event, []).append("added")

        hooks[event] = event_hooks

    settings["hooks"] = hooks

    if not dry_run:
        # Backup existing settings
        if CLAUDE_SETTINGS.exists():
            ts = datetime.now().strftime("%Y%m%d_%H%M%S")
            backup = CLAUDE_SETTINGS.with_suffix(f".bak.{ts}")
            shutil.copy2(CLAUDE_SETTINGS, backup)

        CLAUDE_SETTINGS.parent.mkdir(parents=True, exist_ok=True)
        CLAUDE_SETTINGS.write_text(json.dumps(settings, indent=2) + "\n")

    return actions


def uninstall_hooks() -> int:
    """Remove all Grove hooks from Claude Code settings. Returns count removed."""
    if not CLAUDE_SETTINGS.exists():
        return 0

    try:
        settings = json.loads(CLAUDE_SETTINGS.read_text())
    except (json.JSONDecodeError, OSError):
        return 0

    hooks = settings.get("hooks", {})
    removed = 0

    for event in list(hooks.keys()):
        event_hooks = hooks[event]
        if not isinstance(event_hooks, list):
            continue

        def _is_grove_rule(rule: dict) -> bool:
            # Nested format: {"matcher": "", "hooks": [...]}
            if any(_is_grove_hook(h) for h in rule.get("hooks", [])):
                return True
            # Legacy flat format: {"type": "command", "command": "..."}
            return "command" in rule and "hooks" not in rule and _is_grove_hook(rule)

        filtered = [
            rule for rule in event_hooks if not (isinstance(rule, dict) and _is_grove_rule(rule))
        ]
        removed += len(event_hooks) - len(filtered)

        if filtered:
            hooks[event] = filtered
        else:
            del hooks[event]

    if removed > 0:
        if hooks:
            settings["hooks"] = hooks
        else:
            del settings["hooks"]
        CLAUDE_SETTINGS.write_text(json.dumps(settings, indent=2) + "\n")

    return removed


def is_installed() -> bool:
    """Check if Grove hooks are installed."""
    if not CLAUDE_SETTINGS.exists():
        return False
    try:
        settings = json.loads(CLAUDE_SETTINGS.read_text())
    except (json.JSONDecodeError, OSError):
        return False

    hooks = settings.get("hooks", {})
    for event_hooks in hooks.values():
        if isinstance(event_hooks, list):
            for rule in event_hooks:
                if isinstance(rule, dict):
                    for h in rule.get("hooks", []):
                        if isinstance(h, dict) and _is_grove_hook(h):
                            return True
    return False

File: internal/hook/installer.go
package hook

import (
	"encoding/json"
	"fmt"
	"io"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"time"
)

// HookEvents lists all Claude Code hook event types.
var HookEvents = []string{
	"PreToolUse",
	"PostToolUse",
	"PostToolUseFailure",
	"Stop",
	"SessionStart",
	"SessionEnd",
	"Notification",
	"PermissionRequest",
	"UserPromptSubmit",
	"SubagentStart",
	"SubagentStop",
	"PreCompact",
	"TaskCompleted",
}

const groveMarker = "gw _hook"

// Installer manages Claude Code hook registration in settings.json.
type Installer struct {
	SettingsPath string
	NowFn        func() time.Time
}

// NewInstaller creates an Installer pointing at ~/.claude/settings.json.
func NewInstaller() *Installer {
	home, _ := os.UserHomeDir()
	return &Installer{
		SettingsPath: filepath.Join(home, ".claude", "settings.json"),
		NowFn:        time.Now,
	}
}

// ResolveGW finds the gw binary path.
func ResolveGW() (string, error) {
	path, err := exec.LookPath("gw")
	if err != nil {
		return "", fmt.Errorf("gw not found on PATH; install Grove first")
	}
	return path, nil
}

// IsInstalled checks if Grove hooks are present in settings.json.
func (inst *Installer) IsInstalled() bool {
	settings, err := inst.loadSettings()
	if err != nil {
		return false
	}

	hooks, ok := settings["hooks"].(map[string]any)
	if !ok {
		return false
	}

	for _, eventHooks := range hooks {
		rules, ok := eventHooks.([]any)
		if !ok {
			continue
		}
		for _, rule := range rules {
			if inst.ruleIsGrove(rule) {
				return true
			}
		}
	}
	return false
}

// Install registers Grove hooks in settings.json.
// Returns the number of hooks installed and any error.
func (inst *Installer) Install(gwPath string) (int, error) {
	settings, err := inst.loadSettings()
	if err != nil {
		settings = map[string]any{}
	}

	hooks, ok := settings["hooks"].(map[string]any)
	if !ok {
		hooks = map[string]any{}
	}

	count := 0
	for _, event := range HookEvents {
		command := fmt.Sprintf("GROVE_EVENT=%s %s _hook --event %s", event, gwPath, event)
		ruleEntry := map[string]any{
			"matcher": "",
			"hooks": []any{
				map[string]any{
					"type":    "command",
					"command": command,
				},
			},
		}

		eventHooks, _ := hooks[event].([]any)

		// Find existing grove rule
		groveIdx := -1
		for i, rule := range eventHooks {
			if inst.ruleIsGrove(rule) {
				groveIdx = i
				break
			}
		}

		if groveIdx >= 0 {
			eventHooks[groveIdx] = ruleEntry
		} else {
			eventHooks = append(eventHooks, ruleEntry)
			count++
		}

		hooks[event] = eventHooks
	}

	settings["hooks"] = hooks

	if err := inst.saveSettings(settings); err != nil {
		return 0, err
	}

	if count == 0 {
		count = len(HookEvents) // all updated
	}
	return count, nil
}

// Uninstall removes all Grove hooks from settings.json.
// Returns the number of hooks removed.
func (inst *Installer) Uninstall() (int, error) {
	settings, err := inst.loadSettings()
	if err != nil {
		return 0, nil
	}

	hooks, ok := settings["hooks"].(map[string]any)
	if !ok {
		return 0, nil
	}

	removed := 0
	for event, eventHooksRaw := range hooks {
		eventHooks, ok := eventHooksRaw.([]any)
		if !ok {
			continue
		}

		var filtered []any
		for _, rule := range eventHooks {
			if inst.ruleIsGrove(rule) {
				removed++
			} else {
				filtered = append(filtered, rule)
			}
		}

		if len(filtered) > 0 {
			hooks[event] = filtered
		} else {
			delete(hooks, event)
		}
	}

	if len(hooks) > 0 {
		settings["hooks"] = hooks
	} else {
		delete(settings, "hooks")
	}

	if removed > 0 {
		if err := inst.saveSettings(settings); err != nil {
			return 0, err
		}
	}
	return removed, nil
}

// Backup creates a timestamped backup of settings.json.
// Returns the backup path, or empty string if no settings file exists.
func (inst *Installer) Backup() (string, error) {
	if _, err := os.Stat(inst.SettingsPath); os.IsNotExist(err) {
		return "", nil
	}

	ts := inst.NowFn().Format("20060102_150405")
	backupPath := inst.SettingsPath + ".bak." + ts

	src, err := os.Open(inst.SettingsPath)
	if err != nil {
		return "", fmt.Errorf("opening settings: %w", err)
	}
	defer src.Close()

	dst, err := os.Create(backupPath)
	if err != nil {
		return "", fmt.Errorf("creating backup: %w", err)
	}
	defer dst.Close()

	if _, err := io.Copy(dst, src); err != nil {
		return "", fmt.Errorf("copying settings: %w", err)
	}

	return backupPath, nil
}

func (inst *Installer) ruleIsGrove(rule any) bool {
	ruleMap, ok := rule.(map[string]any)
	if !ok {
		return false
	}
	innerHooks, ok := ruleMap["hooks"].([]any)
	if !ok {
		return false
	}
	for _, h := range innerHooks {
		hMap, ok := h.(map[string]any)
		if !ok {
			continue
		}
		cmd, _ := hMap["command"].(string)
		if strings.Contains(cmd, groveMarker) {
			return true
		}
	}
	return false
}

func (inst *Installer) loadSettings() (map[string]any, error) {
	data, err := os.ReadFile(inst.SettingsPath)
	if err != nil {
		return nil, err
	}
	var settings map[string]any
	if err := json.Unmarshal(data, &settings); err != nil {
		return nil, fmt.Errorf("parsing settings.json: %w", err)
	}
	return settings, nil
}

func (inst *Installer) saveSettings(settings map[string]any) error {
	data, err := json.MarshalIndent(settings, "", "  ")
	if err != nil {
		return err
	}
	data = append(data, '\n')

	if err := os.MkdirAll(filepath.Dir(inst.SettingsPath), 0o755); err != nil {
		return err
	}

	tmp := inst.SettingsPath + ".tmp"
	if err := os.WriteFile(tmp, data, 0o644); err != nil {
		return err
	}
	return os.Rename(tmp, inst.SettingsPath)
}
